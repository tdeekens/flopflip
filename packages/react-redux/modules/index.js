const version = VERSION;

export {
  createFlopFlipEnhancer,
  STATE_SLICE as FLOPFLIP_STATE_SLICE,
} from './store';
export {
  flopflipReducer,
  updateStatus,
  updateFlags,
  UPDATE_STATUS,
  UPDATE_FLAGS,
} from './ducks';
export {
  FeatureToggled,
  injectFeatureToggle,
  injectFeatureToggles,
  branchOnFeatureToggle,
  ConfigureFlopFlip,
} from './components';

export { version };
