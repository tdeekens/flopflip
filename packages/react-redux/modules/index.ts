export {
  createFlopFlipEnhancer,
  STATE_SLICE as FLOPFLIP_STATE_SLICE,
} from './store';
export {
  createFlopflipReducer,
  flopflipReducer,
  selectFlags as selectFeatureFlags,
  selectFlag as selectFeatureFlag,
  UPDATE_STATUS,
  UPDATE_FLAGS,
} from './ducks';
export {
  ToggleFeature,
  injectFeatureToggle,
  injectFeatureToggles,
  branchOnFeatureToggle,
  ConfigureFlopFlip,
  ReconfigureFlopFlip,
} from './components';

export { version } from '../package.json';
