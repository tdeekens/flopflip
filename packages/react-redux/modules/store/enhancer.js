import { initialize, listen } from '@flopflip/launchdarkly-wrapper';
import { updateFlags, updateStatus } from './../ducks';

export default function createFlopFlipEnhancer(clientSideId, user) {
  const client = initialize({ clientSideId, user });

  return next => (...args) => {
    const store = next(...args);

    listen({
      client,
      onUpdateFlags: store.dispatch(updateFlags),
      onUpdateStatus: store.dispatch(updateStatus),
    });

    return store;
  };
}
