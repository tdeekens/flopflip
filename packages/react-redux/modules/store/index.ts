export { default as createFlopFlipEnhancer } from './enhancer';
export const STATE_SLICE = '@flopflip';
