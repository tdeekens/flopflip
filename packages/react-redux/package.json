{
  "name": "@flopflip/react-redux",
  "version": "2.0.1",
  "description":
    "A feature toggle wrapper to use LaunchDarkly with React Redux",
  "main": "dist/@flopflip-react-redux.cjs.js",
  "module": "dist/@flopflip-react-redux.es.js",
  "browser": "dist/@flopflip-react-redux.umd.js",
  "scripts": {
    "preversion": "npm run build",
    "prebuild": "rimraf dist/**",
    "build": "cross-env run-p build:umd build:umd:min build:es build:cjs",
    "build:watch": "cross-env npm run build:es -- -w",
    "build:umd":
      "cross-env NODE_ENV=development BABEL_ENV=rollup rollup -c -f umd -o dist/@flopflip-react-redux.umd.js",
    "build:umd:min":
      "cross-env NODE_ENV=production BABEL_ENV=rollup rollup -c -f umd -o dist/@flopflip-react-redux.umd.min.js",
    "build:es":
      "cross-env NODE_ENV=development BABEL_ENV=rollup rollup -c -f es -o dist/@flopflip-react-redux.es.js",
    "build:cjs":
      "cross-env NODE_ENV=development BABEL_ENV=rollup rollup -c -f cjs -o dist/@flopflip-react-redux.cjs.js"
  },
  "files": ["readme.md", "dist/**"],
  "publishConfig": {
    "access": "public"
  },
  "repository": {
    "type": "git",
    "url": "git+https://github.com/tdeekens/flopflip.git"
  },
  "author": "Tobias Deekens <nerd@tdeekens.name>",
  "license": "MIT",
  "bugs": {
    "url": "https://github.com/tdeekens/flopflip/issues"
  },
  "homepage": "https://github.com/tdeekens/flopflip#readme",
  "devDependencies": {
    "enzyme": "^2.9.1",
    "enzyme-to-json": "^1.5.1",
    "prop-types": "15.5.10",
    "react": "^15.6.1",
    "react-dom": "^15.6.1",
    "react-redux": "^5.0.6",
    "react-test-renderer": "^15.6.1",
    "redux": "^3.7.2"
  },
  "dependencies": {
    "@flopflip/launchdarkly-wrapper": "^2.1.1",
    "@flopflip/react": "^2.0.1",
    "recompose": "^0.25.0"
  },
  "peerDependencies": {
    "prop-types": "15.5.10",
    "react": "^15.6.0",
    "react-dom": "^15.6.0",
    "react-redux": "^5.0.5",
    "redux": "^3.7.2"
  },
  "keywords": ["feature-flags", "feature-toggles", "LaunchDarkly", "client"]
}
