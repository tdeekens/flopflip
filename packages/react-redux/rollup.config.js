const resolve = require('rollup-plugin-node-resolve');
const commonjs = require('rollup-plugin-commonjs');
const babel = require('rollup-plugin-babel');
const replace = require('rollup-plugin-replace');
const uglify = require('rollup-plugin-uglify');
const builtins = require('rollup-plugin-node-builtins');

const env = process.env.NODE_ENV;
const version = process.env.npm_package_version;

const config = {
  input: 'modules/index.js',
  name: '@flopflip-react-redux',
  sourcemap: true,
  external: ['react', 'prop-types', 'redux', 'react-redux'],
  globals: {
    react: 'React',
    'prop-types': 'PropTypes',
    redux: 'redux',
    'react-redux': 'react-redux',
  },
  plugins: [
    commonjs(),
    babel({
      babelrc: false,
      exclude: 'node_modules/**',
      presets: ['es2015-rollup', 'stage-0', 'react'],
      plugins: ['transform-react-remove-prop-types'],
    }),
    replace({
      'process.env.NODE_ENV': JSON.stringify(env),
      VERSION: `'${version}'`,
    }),
    resolve({
      module: true,
    }),
    builtins(),
  ],
};

if (env === 'production') {
  config.plugins.push(
    uglify({
      compress: {
        warnings: false,
      },
    })
  );
}

module.exports = config;
