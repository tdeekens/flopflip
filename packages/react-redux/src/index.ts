const version = '__@FLOPFLIP/VERSION_OF_RELEASE__';

export { default as createFlopFlipEnhancer } from './store/enhancer';
// Import this separately to avoid a circular dependency
export {
  ConfigureFlopFlip,
  ReconfigureFlopFlip,
  ToggleFeature,
} from './components';
export {
  createFlopflipReducer,
  flopflipReducer,
  selectFlag as selectFeatureFlag,
  selectFlags as selectFeatureFlags,
  UPDATE_FLAGS,
  UPDATE_STATUS,
} from './ducks';
export {
  branchOnFeatureToggle,
  injectFeatureToggle,
  injectFeatureToggles,
} from './hocs';
export {
  useAdapterReconfiguration,
  useAdapterStatus,
  useFeatureToggle,
  useFeatureToggles,
  useFlagVariation,
  useFlagVariations,
} from './hooks';
export { STATE_SLICE as FLOPFLIP_STATE_SLICE } from './store/constants';

export { version };
