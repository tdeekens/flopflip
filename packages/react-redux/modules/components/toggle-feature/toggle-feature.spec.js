import { mount } from 'enzyme';
import { Provider } from 'react-redux';
import React from 'react';
import memoryAdapter from '@flopflip/memory-adapter';
import configureStore from 'redux-mock-store';
import Configure from '../configure';
import { STATE_SLICE } from './../../store';
import ToggleFeature, { mapStateToProps } from './toggle-feature';

describe('mapStateToProps', () => {
  describe('with existing `flag` ', () => {
    let state;

    beforeEach(() => {
      state = {
        [STATE_SLICE]: { flags: { flag1: true } },
      };
    });

    it('should map `isFeatureEnabled` as `true` onto `props`', () => {
      expect(mapStateToProps(state, { flag: 'flag1' }).isFeatureEnabled).toBe(
        true
      );
    });
  });

  describe('without existing `flag` ', () => {
    let state;

    beforeEach(() => {
      state = {
        [STATE_SLICE]: { flags: {} },
      };
    });

    it('should map `isFeatureEnabled` as `false` onto `props`', () => {
      expect(mapStateToProps(state, { flag: 'flag1' }).isFeatureEnabled).toBe(
        false
      );
    });
  });

  describe('with matching `variation`', () => {
    let state;

    beforeEach(() => {
      state = {
        [STATE_SLICE]: { flags: { flag1: 'flagVariate1' } },
      };
    });

    it('should map `isFeatureEnabled` as `true` onto `props`', () => {
      expect(
        mapStateToProps(state, {
          flag: 'flag1',
          variation: 'flagVariate1',
        }).isFeatureEnabled
      ).toBe(true);
    });
  });

  describe('with non-matching `variation`', () => {
    let state;

    beforeEach(() => {
      state = {
        [STATE_SLICE]: { flags: { flag1: 'flagVariate1' } },
      };
    });

    it('should map `isFeatureEnabled` as `true` onto `props`', () => {
      expect(
        mapStateToProps(state, {
          flag: 'flag1',
          variation: 'flagVariate2',
        }).isFeatureEnabled
      ).toBe(false);
    });
  });
});

const FeatureComponent = () => <div />;
FeatureComponent.displayName = 'FeatureComponent';

const createTestProps = custom => ({
  adapterArgs: {},

  ...custom,
});
const createMockStore = configureStore();

describe('<ToggleFeature>', () => {
  describe('when feature is disabled', () => {
    let props;
    let wrapper;
    beforeEach(() => {
      props = createTestProps();
      const store = createMockStore({
        [STATE_SLICE]: { flags: { flag1: false } },
      });
      props = createTestProps();
      wrapper = mount(
        <Provider store={store}>
          <Configure {...props} adapter={memoryAdapter}>
            <ToggleFeature flag="flag1">
              <FeatureComponent />
            </ToggleFeature>
          </Configure>
        </Provider>
      );
    });

    it('should match snapshot', () => {
      expect(wrapper).toMatchSnapshot();
    });

    it('should not render the `FeatureComponent`', () => {
      expect(wrapper).not.toRender(FeatureComponent);
    });

    describe('when enabling feature', () => {
      beforeEach(() => {
        props = createTestProps();
        const store = createMockStore({
          [STATE_SLICE]: { flags: { flag1: true } },
        });
        props = createTestProps();
        wrapper = mount(
          <Provider store={store}>
            <Configure {...props} adapter={memoryAdapter}>
              <ToggleFeature flag="flag1">
                <FeatureComponent />
              </ToggleFeature>
            </Configure>
          </Provider>
        );
      });

      it('should match snapshot', () => {
        expect(wrapper).toMatchSnapshot();
      });

      it('should not render the `FeatureComponent`', () => {
        expect(wrapper).toRender(FeatureComponent);
      });
    });
  });
});
