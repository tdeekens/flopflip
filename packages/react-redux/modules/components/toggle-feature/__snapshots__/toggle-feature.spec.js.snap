// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ToggleFeature> when feature is disabled should match snapshot 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <Connect(ConfigureFlopflip)
    adapter={
      Object {
        "configure": [Function],
        "getFlag": [Function],
        "getIsReady": [Function],
        "reconfigure": [Function],
        "reset": [Function],
        "waitUntilConfigured": [Function],
      }
    }
    adapterArgs={Object {}}
  >
    <ConfigureFlopflip
      adapter={
        Object {
          "configure": [Function],
          "getFlag": [Function],
          "getIsReady": [Function],
          "reconfigure": [Function],
          "reset": [Function],
          "waitUntilConfigured": [Function],
        }
      }
      adapterArgs={Object {}}
      defaultFlags={Object {}}
      handleUpdateFlags={[Function]}
      handleUpdateStatus={[Function]}
      shouldDeferAdapterConfiguration={false}
    >
      <ConfigureAdapter
        adapter={
          Object {
            "configure": [Function],
            "getFlag": [Function],
            "getIsReady": [Function],
            "reconfigure": [Function],
            "reset": [Function],
            "waitUntilConfigured": [Function],
          }
        }
        adapterArgs={
          Object {
            "onFlagsStateChange": [Function],
            "onStatusStateChange": [Function],
          }
        }
        defaultFlags={Object {}}
        render={null}
        shouldDeferAdapterConfiguration={false}
      >
        <ToggleFeature
          flag="flag1"
        >
          <ToggleFeature
            dispatch={[Function]}
            flag="flag1"
            isFeatureEnabled={false}
            render={null}
            toggledComponent={null}
            untoggledComponent={null}
          />
        </ToggleFeature>
      </ConfigureAdapter>
    </ConfigureFlopflip>
  </Connect(ConfigureFlopflip)>
</Provider>
`;

exports[`<ToggleFeature> when feature is disabled when enabling feature should match snapshot 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <Connect(ConfigureFlopflip)
    adapter={
      Object {
        "configure": [Function],
        "getFlag": [Function],
        "getIsReady": [Function],
        "reconfigure": [Function],
        "reset": [Function],
        "waitUntilConfigured": [Function],
      }
    }
    adapterArgs={Object {}}
  >
    <ConfigureFlopflip
      adapter={
        Object {
          "configure": [Function],
          "getFlag": [Function],
          "getIsReady": [Function],
          "reconfigure": [Function],
          "reset": [Function],
          "waitUntilConfigured": [Function],
        }
      }
      adapterArgs={Object {}}
      defaultFlags={Object {}}
      handleUpdateFlags={[Function]}
      handleUpdateStatus={[Function]}
      shouldDeferAdapterConfiguration={false}
    >
      <ConfigureAdapter
        adapter={
          Object {
            "configure": [Function],
            "getFlag": [Function],
            "getIsReady": [Function],
            "reconfigure": [Function],
            "reset": [Function],
            "waitUntilConfigured": [Function],
          }
        }
        adapterArgs={
          Object {
            "onFlagsStateChange": [Function],
            "onStatusStateChange": [Function],
          }
        }
        defaultFlags={Object {}}
        render={null}
        shouldDeferAdapterConfiguration={false}
      >
        <ToggleFeature
          flag="flag1"
        >
          <ToggleFeature
            dispatch={[Function]}
            flag="flag1"
            isFeatureEnabled={true}
            render={null}
            toggledComponent={null}
            untoggledComponent={null}
          >
            <FeatureComponent>
              <div />
            </FeatureComponent>
          </ToggleFeature>
        </ToggleFeature>
      </ConfigureAdapter>
    </ConfigureFlopflip>
  </Connect(ConfigureFlopflip)>
</Provider>
`;
